File: workers/python/src/variables/variable.py
import asyncio
from typing import Any

from variables.constants import VariableMode, VariableType
from variables.variable_client import VariableClient


async def build_variables(
    variable_spec: dict[str, dict], variable_client
) -> dict[str, Any]:
    def is_object(thing):
        return thing is not None and isinstance(thing, dict)

    if not is_object(variable_spec):
        raise ValueError("Failed to build the variable: invalid variableSpec.")

    ret = {}
    variables_read_in_advance = []
    for variable_name, variable_property in variable_spec.items():
        if not is_object(variable_property):
            raise ValueError(f"Failed to build the variable: {variable_name}")

        key, type, mode = (
            variable_property.get("key"),
            variable_property.get("type"),
            variable_property.get("mode"),
        )

        allow_write = mode == VariableMode.READ_WRITE

        if type == VariableType.SIMPLE or type == VariableType.NATIVE:
            variables_read_in_advance.append(
                {"variable_name": variable_name, "key": key, "type": type, "mode": mode}
            )
        elif type == VariableType.ADVANCED:
            ret[variable_name] = AdvancedVariable(key, variable_client, allow_write)  # type: ignore

    values = await variable_client.readMany(
        [v["key"] for v in variables_read_in_advance]
    )

    for i, variable in enumerate(variables_read_in_advance):
        variable_name, key, mode, type = (
            variable["variable_name"],
            variable["key"],
            variable["mode"],
            variable["type"],
        )
        value = values[i]

        allow_write = mode == VariableMode.READ_WRITE

        if type == VariableType.SIMPLE:
            ret[variable_name] = SimpleVariable(
                key, value, variable_client, allow_write
            )  # type: ignore
        elif type == VariableType.NATIVE:
            ret[variable_name] = value
        else:
            raise Exception(
                "This should not happen: pre-read variables are not simple or native."
            )

    return ret


class SimpleVariable:
    def __init__(
        self,
        store_key: str,
        initial_value: Any,
        variable_client: VariableClient,
        allow_write: bool,
    ):
        self.store_key = store_key
        self.variable_client = variable_client
        self.value = initial_value
        self.allow_write = allow_write

    # def get(self):
    #     return self.value

    def set(self, value: Any):
        if not self.allow_write:
            raise Exception("Variable write is forbidden.")
        self.value = value
        self.variable_client.writeBuffer(self.store_key, value)


class AdvancedVariable:
    def __init__(
        self, store_key: str, variable_client: VariableClient, allow_write: bool
    ):
        self.store_key = store_key
        self.variable_client = variable_client
        self.allow_write = allow_write

    def get(self):
        result = asyncio.run(self.variable_client.read(self.store_key))
        return result

    def set(self, value: Any):
        if not self.allow_write:
            raise Exception("Variable write is forbidden.")
        asyncio.run(self.variable_client.write(self.store_key, value))


# def variablesCodeGen(variablesSpec: dict):
#     for scope, variables in variablesSpec.items():
#

File: workers/javascript/packages/shared/src/plugins/execution/variable.ts
import { toVmValue, hostFunction } from '@superblocks/wasm-sandbox-js';
import { decodeBytestrings } from '../../utils';
import { VariableClient } from './utils';

export enum VariableType {
  Simple = 'TYPE_SIMPLE',
  Advanced = 'TYPE_ADVANCED',
  Native = 'TYPE_NATIVE'
}

export enum VariableMode {
  ReadWrite = 'MODE_READWRITE',
  Read = 'MODE_READ'
}

export const buildVariables = async (
  variableSpec: Record<string, { key: string; type: string; mode?: string }>,
  variableClient: VariableClient
): Promise<{ [key: string]: unknown }> => {
  const isObject = (thing) => {
    return thing !== null && thing !== undefined && typeof thing === 'object';
  };

  if (!isObject(variableSpec)) {
    throw new Error('Failed to build the variable: invalid variableSpec.');
  }

  const ret: Record<string, unknown> = {};
  const variablesReadInAdvance: Array<{ variableName: string; key: string; mode?: string; type: string }> = [];
  for (const [variableName, variableProperty] of Object.entries(variableSpec)) {
    if (!isObject(variableProperty)) {
      throw new Error(`Failed to build the variable: ${variableName}`);
    }

    const { key, type, mode } = variableProperty;

    const allowWrite = mode === VariableMode.ReadWrite;

    if (type === VariableType.Simple || type === VariableType.Native) {
      variablesReadInAdvance.push({ variableName, ...variableProperty });
    } else if (type === VariableType.Advanced) {
      ret[variableName] = new AdvancedVariable(key, variableClient, allowWrite);
    }
  }

  const values = (await variableClient.read(variablesReadInAdvance.map((variable) => variable.key))).data;

  for (let i = 0; i < variablesReadInAdvance.length; i++) {
    const { variableName, key, mode, type } = variablesReadInAdvance[i];
    const value = values[i];

    const allowWrite = mode === VariableMode.ReadWrite;

    if (type === VariableType.Simple) {
      ret[variableName] = new SimpleVariable(key, value, variableClient, allowWrite);
    } else {
      // adapt to the buffer type output
      if ((value as { output: unknown })?.output) {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        (value as any).output = decodeBytestrings((value as any).output, true);
      }
      ret[variableName] = value;
    }
  }

  return ret;
};

export class SimpleVariable {
  key;
  value;
  private variableClient;
  private allowWrite;

  constructor(key: string, initialValue: unknown, variableClient: VariableClient, allowWrite: boolean) {
    this.key = key;
    this.value = initialValue;
    this.variableClient = variableClient;
    this.allowWrite = allowWrite;
  }

  set(value: unknown): void {
    if (!this.allowWrite) {
      throw new Error('Write is not allowed.');
    }

    this.value = value;
    this.variableClient.writeBuffer(this.key, this.value);
  }

  [toVmValue]() {
    return {
      key: this.key,
      value: this.value,
      set: hostFunction(this.set.bind(this))
    };
  }
}

export class AdvancedVariable {
  key;
  private variableClient;
  private allowWrite;

  constructor(key: string, variableClient: VariableClient, allowWrite: boolean) {
    this.key = key;
    this.variableClient = variableClient;
    this.allowWrite = allowWrite;
  }

  async get(): Promise<unknown> {
    const raw = (await this.variableClient.read([this.key])).data;
    if (!raw || !raw.length) {
      throw new Error('Variable does not exist.');
    }

    try {
      return raw[0];
    } catch (e) {
      throw new Error('failed to parse the store value into object');
    }
  }

  async set(value: unknown): Promise<void> {
    if (!this.allowWrite) {
      throw new Error('Write is not allowed.');
    }

    try {
      await this.variableClient.write(this.key, value);
    } catch (e) {
      throw new Error(`failed to write to variable store: ${e}`);
    }
  }

  [toVmValue]() {
    return {
      key: this.key,
      get: hostFunction(this.get.bind(this)),
      set: hostFunction(this.set.bind(this))
    };
  }
}
